### ISSUE TITLE
[RyuJIT] TailcallStress needs to check more conditions

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#11408]
When `COMPlus_TailcallStress=1` is set, the JIT converts as many call/ret pairs as possible into tail call, which it treats as `tail.` prefixed "explicit" tail calls. It does not in this case run the legality checks that are run when checking to determine if implicit tailcall optimization can be done, specifically the code under `FEATURE_TAILCALL_OPT` in fgMorphCall() that rejects a potential tail call if any variable in the function has had its address taken, if there are pinned locals, or if there are struct promoted struct arguments.

This causes code such as this to fail:
```
    unsafe class Program
    {
        static void foo(IntPtr x)
        {
            Console.WriteLine(*(int*)x);
        }
        static void Main(string[] args)
        {
            int x = 42;
            foo((IntPtr)(&x));
        }
    }
```

**Comment by leppie** [comment]
A tail call will implicitly discard the stack, so I am not sure how this can be a viable candidate. 

PS: Also this is just a tail call, not a tail call optimization. This will in fact run slower than a normal direct call in the current JIT. TCO is an ambiguous term that should not be used.

**Comment by BruceForstall** [comment]
@leppie yes, the JIT bug is that in this particular stress mode, the JIT does not do enough legality checking, therefore makes an illegal tail call.

**Comment by BruceForstall** [comment]
(Internal note: this came from VSO#278840)

**Comment by BruceForstall** [comment]
If/when fixed, enable the test added with #11410

**Comment by BruceForstall** [comment]
@jakobbotsch @jashook fyi

**Comment by jakobbotsch** [comment]
IMO it does not make sense to spend time on trying to make tailcall stress work -- instead ABI stress should be expanded to cover the missing cases (generating tests that are correct by construction).
